ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Server.scala

```diff
@@ -0,0 +1,395 @@
+      (remotehost, remoteport) = (address.host, address.port)
+      ref <- Resource.eval(Concurrent[F].ref(Map[Int, H2Stream[F]]()))
+      initialWriteBlock <- Resource.eval(Deferred[F, Either[Throwable, Unit]])
+      stateRef <- Resource.eval(
+        Concurrent[F].ref(
+          H2Connection.State(
+            initialRemoteSettings,
+            defaultSettings.initialWindowSize.windowSize,
+            initialWriteBlock,
+            localSettings.initialWindowSize.windowSize,
+            0,
+            0,
+            false,
+            None,
+            None,
+          )
+        )
+      )
+      queue <- Resource.eval(cats.effect.std.Queue.unbounded[F, Chunk[H2Frame]]) // TODO revisit
+      hpack <- Resource.eval(Hpack.create[F])
+      settingsAck <- Resource.eval(
+        Deferred[F, Either[Throwable, H2Frame.Settings.ConnectionSettings]]
+      )
+      streamCreationLock <- Resource.eval(cats.effect.std.Semaphore[F](1))
+      // data <- Resource.eval(cats.effect.std.Queue.unbounded[F, Frame.Data])
+      created <- Resource.eval(cats.effect.std.Queue.unbounded[F, Int])
+      closed <- Resource.eval(cats.effect.std.Queue.unbounded[F, Int])
+
+      h2 = new H2Connection(
+        remotehost,
+        remoteport,
+        H2Connection.ConnectionType.Server,
+        localSettings,
+        ref,
+        stateRef,
+        queue,
+        created,
+        closed,
+        hpack,
+        streamCreationLock.permit,
+        settingsAck,
+        ByteVector.empty,
+        socket,
+        logger,
+      )
+      bgWrite <- h2.writeLoop.compile.drain.background
+      _ <- Resource.eval(
+        queue.offer(Chunk.singleton(H2Frame.Settings.ConnectionSettings.toSettings(localSettings)))
+      )
+      bgRead <- h2.readLoop.compile.drain.background
+
+      settings <- Resource.eval(h2.settingsAck.get.rethrow)
+      // h2c Initial Request Communication on h2c Upgrade
+      _ <- Resource.eval(
+        initialRequest.traverse_(req =>
+          h2.initiateRemoteStreamById(1)
+            .flatMap(h2Stream =>
+              h2Stream.state
+                .modify { s =>
+                  val x = s.copy(state = H2Stream.StreamState.HalfClosedRemote)
+                  (x, x)
+                }
+                .flatMap(s =>
+                  s.request.complete(Either.right(req)) >>
+                    s.readBuffer.offer(
+                      Either
+                        .right(req.body.compile.to(fs2.Collector.supportsByteVector(ByteVector)))
+                    ) >>
+                    s.writeBlock.complete(Either.right(()))
+                ) >> created.offer(1)
+            )
+        )
+      )
+      _ <-
+        Stream
+          .fromQueueUnterminated(closed)
+          .map(i =>
+            Stream.eval(
+              // Max Time After Close We Will Still Accept Messages
+              (Temporal[F].sleep(1.seconds) >>
+                ref.update(m => m - i)).timeout(15.seconds).attempt.start
```

Comment:
Are these magic numbers holy or should they be parameterized/

Reply:
They could be parameterized. This is mostly a background system designed to eventually remove dead streams. It could be parameterized if we really wanted to but it would extend our configuration space considerably.